mne/time_frequency/psd.py

```diff
@@ -112,12 +111,209 @@ def _check_nfft(n, n_fft, n_overlap):
+        it is the smoother are the PSDs. The default value is 256.
+        If ``n_fft > len(epochs.times)``, it will be adjusted down to
+        ``len(epochs.times)``.
+    n_overlap : int
+        The number of points of overlap between blocks. Will be adjusted
+        to be <= n_fft.
+        bandwidth : float
+        The bandwidth of the multi taper windowing function in Hz.
+    picks : array-like of int | None
+        The selection of channels to include in the computation.
+        If None, take all channels.
+    proj : bool
+        Apply SSP projection vectors. If inst is ndarray this is not used.
+    sfreq : float
+        The sampling frequency of the data. Required if inst is an array,
+        otherwise this is not used and sfreq is pulled from inst.
+    n_jobs : int
+        Number of CPUs to use in the computation.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see mne.verbose).
+
+    Returns
+    -------
+    psds : ndarray, shape ([n_epochs], n_channels, n_freqs)
+        The power spectral densities. If Raw is provided,
+        then psds will be 2-D.
+    freqs : ndarray (n_freqs)
+        The frequencies.
+    """
+    from scipy.signal import welch
+
+    # Prep data
+    data, sfreq = _check_psd_data(inst, tmin, tmax, picks, proj, sfreq)
+    dshape = data.shape[:-1]
+    n_times = data.shape[-1]
+
+    # This will return the same object if len(dshape) == 1
+    data = data.reshape(np.product(dshape), n_times)
+
+    # Prep the PSD
+    n_fft, n_overlap = _check_nfft(n_times, n_fft, n_overlap)
+    win_size = n_fft / float(sfreq)
+    logger.info("Effective window size : %0.3f (s)" % win_size)
+    freqs = np.arange(n_fft // 2 + 1, dtype=float) * (sfreq / n_fft)
+    freq_mask = (freqs >= fmin) & (freqs <= fmax)
+    freqs = freqs[freq_mask]
+    n_freqs = len(freqs)
+
+    # Parallelize across first N-1 dimensions
+    psds = np.empty(data.shape[:-1] + (freqs.size,))
+    parallel, my_pwelch, n_jobs = parallel_func(_pwelch, n_jobs=n_jobs,
+                                                verbose=verbose)
+    data_splits = np.array_split(data, n_jobs)
+    f_psd = parallel(my_pwelch(d, noverlap=n_overlap, nfft=n_fft,
+                     fs=sfreq, freq_mask=freq_mask,
+                     welch_fun=welch)
+                     for d in data_splits)
+
+    # Combining/reshaping to original data shape
+    psds = np.concatenate(f_psd, axis=0)
+    psds = psds.reshape(np.hstack([dshape, n_freqs]))
+    return psds, freqs
+
+
+@verbose
+def psd_multitaper(inst, fmin=0, fmax=np.inf, tmin=None, tmax=None,
+                   bandwidth=None, adaptive=False, low_bias=True,
+                   normalization='length', picks=None, proj=False,
+                   sfreq=None, n_jobs=1, verbose=None):
+    """Compute the PSD using multitapers.
+
+
+    Parameters
+    ----------
+    inst : instance of Epochs or Raw |
+           array of shape ([n_epochs], n_channels, n_times)
```

Comment:
sounds good, will update in a bit

On Sat, Dec 26, 2015 at 9:31 PM, Mainak Jas <notifications@github.com>
wrote:

> In mne/time_frequency/psd.py
> <https://github.com/mne-tools/mne-python/pull/2710#discussion_r48451451>:
>
> > +    psds = psds.reshape(np.hstack([dshape, n_freqs]))
> > +    return psds, freqs
> > +
> > +
> > +@verbose
> > +def psd_multitaper(inst, fmin=0, fmax=np.inf, tmin=None, tmax=None,
> > +                   bandwidth=None, adaptive=False, low_bias=True,
> > +                   normalization='length', picks=None, proj=False,
> > +                   sfreq=None, n_jobs=1, verbose=None):
> > +    """Compute the PSD using multitapers.
> > +
> > +
> > +    Parameters
> > +    ----------
> > +    inst : instance of Epochs or Raw |
> > +           array of shape ([n_epochs], n_channels, n_times)
>
> don't allow array. It's inst. What you do instead is use this pattern:
>
> def _psd_estimator(data, ...):
>     """ Here data = inst of epochs | raw | ndarray but not in public API    """
>     if isinstance(data, Epochs):
>        data = epochs.get_data()
>     # now operate on arrays here
> def psd_multitaper(inst, ...):
>   return _psd_mutitaper(inst, ...)
> class PSDEstimator:
>    def __init__():
>        # blah blah code
>        pass
>    def fit():
>       return _psd_estimator(epochs_data, ...)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/2710/files#r48451451>.
>